File: tests/functional/Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Dialogs/FileUpload/FileUploadDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.IO;
using System.Net;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.FileUpload
{
    public class FileUploadDialog : ComponentDialog
    {
        public FileUploadDialog()
            : base(nameof(FileUploadDialog))
        {
            AddDialog(new AttachmentPrompt(nameof(AttachmentPrompt)));
            AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt)));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { PromptUploadStepAsync, HandleAttachmentStepAsync, FinalStepAsync }));

            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> PromptUploadStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var promptOptions = new PromptOptions
            {
                Prompt = MessageFactory.Text("Please upload a file to continue."),
                RetryPrompt = MessageFactory.Text("You must upload a file."),
            };

            return await stepContext.PromptAsync(nameof(AttachmentPrompt), promptOptions, cancellationToken);
        }

        private async Task<DialogTurnResult> HandleAttachmentStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var fileText = string.Empty;

            foreach (var file in stepContext.Context.Activity.Attachments)
            {
                var remoteFileUrl = file.ContentUrl;
                var localFileName = Path.Combine(Path.GetTempPath(), file.Name);
                string fileContent;

                using (var webClient = new WebClient())
                {
                    webClient.DownloadFile(remoteFileUrl, localFileName);
                    using var reader = new StreamReader(localFileName);
                    fileContent = await reader.ReadToEndAsync();
                }

                fileText += $"Attachment \"{file.Name}\" has been received.\r\n";
                fileText += $"File content: {fileContent}\r\n";
            }

            await stepContext.Context.SendActivityAsync(MessageFactory.Text(fileText), cancellationToken);

            // Ask to upload another file or end.
            const string messageText = "Do you want to upload another file?";
            const string repromptMessageText = "That's an invalid choice.";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput)
            };

            return await stepContext.PromptAsync(nameof(ConfirmPrompt), options, cancellationToken);
        }

        private async Task<DialogTurnResult> FinalStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var tryAnother = (bool)stepContext.Result;
            if (tryAnother)
            {
                return await stepContext.ReplaceDialogAsync(InitialDialogId, cancellationToken: cancellationToken);
            }

            return new DialogTurnResult(DialogTurnStatus.Complete);
        }
    }
}

File: tests/functional/Bots/JavaScript/Skills/CodeFirst/WaterfallSkillBot/dialogs/fileUpload/fileUploadDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { MessageFactory, InputHints } = require('botbuilder');
const { ComponentDialog, AttachmentPrompt, WaterfallDialog, DialogTurnStatus, ConfirmPrompt } = require('botbuilder-dialogs');
const fs = require('fs');
const fetch = require('node-fetch');
const os = require('os');
const path = require('path');
const stream = require('stream');
const util = require('util');

const streamPipeline = util.promisify(stream.pipeline);

const ATTACHMENT_PROMPT = 'AttachmentPrompt';
const CONFIRM_PROMPT = 'ConfirmPrompt';
const WATERFALL_DIALOG = 'WaterfallDialog';

class FileUploadDialog extends ComponentDialog {
  /**
   * @param {string} dialogId
   */
  constructor (dialogId) {
    super(dialogId);

    this.addDialog(new AttachmentPrompt(ATTACHMENT_PROMPT))
      .addDialog(new ConfirmPrompt(CONFIRM_PROMPT))
      .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
        this.promptUploadStep.bind(this),
        this.handleAttachmentStep.bind(this),
        this.finalStep.bind(this)
      ]));

    this.initialDialogId = WATERFALL_DIALOG;
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async promptUploadStep (stepContext) {
    return stepContext.prompt(ATTACHMENT_PROMPT, {
      prompt: MessageFactory.text('Please upload a file to continue.'),
      retryPrompt: MessageFactory.text('You must upload a file.')
    });
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async handleAttachmentStep (stepContext) {
    let fileText = '';
    let fileContent = '';

    for (const file of stepContext.context.activity.attachments) {
      const localFileName = path.resolve(os.tmpdir(), file.name);
      const tempFile = fs.createWriteStream(localFileName);
      const response = await fetch(file.contentUrl);
      await streamPipeline(response.body, tempFile);

      fileContent = fs.readFileSync(localFileName, 'utf8');
      fileText += `Attachment "${file.name}" has been received.\r\n`;
      fileText += `File content: ${fileContent}\r\n`;
    }

    await stepContext.context.sendActivity(MessageFactory.text(fileText));

    // Ask to upload another file or end.
    const messageText = 'Do you want to upload another file?';
    const repromptMessageText = "That's an invalid choice.";

    return stepContext.prompt(CONFIRM_PROMPT, {
      prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
      retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput)
    });
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async finalStep (stepContext) {
    const tryAnother = stepContext.result;

    if (tryAnother) {
      return stepContext.replaceDialog(this.initialDialogId);
    }

    return { status: DialogTurnStatus.complete };
  }
}

module.exports.FileUploadDialog = FileUploadDialog;
